File: sdk/kotlin/src/main/kotlin/com/privacyrpc/sdk/NotificationHub.kt
package com.privacyrpc.sdk

/**
 * Notification Hub for PrivacyRPC
 *
 * Central notification management matching the Chrome extension's
 * notification system with throttling, type-based filtering,
 * and multi-channel support.
 */
class NotificationHub(
    private var settings: NotificationSettings = NotificationSettings()
) {

    // ── Notification Types ───────────────────────────────────

    enum class NotificationType(
        val priority: Int,
        val supportsNative: Boolean,
        val supportsOverlay: Boolean
    ) {
        TOR_CONNECTED(80, true, false),
        TOR_DISCONNECTED(80, true, false),
        PROXY_ERROR(100, true, true),
        PROTECTION_ON(50, true, false),
        PROTECTION_OFF(80, true, true),
        SUSPICIOUS_RPC(100, true, true),
        EXT_WARNING(80, true, true),
        UNPROTECTED_DAPP(50, false, true),
        RPC_BLOCKED(80, false, true)
    }

    // ── Settings ─────────────────────────────────────────────

    data class NotificationSettings(
        val nativeEnabled: Boolean = true,
        val overlayEnabled: Boolean = true,
        val native: NativeSettings = NativeSettings(),
        val overlay: OverlaySettings = OverlaySettings(),
        val throttling: ThrottlingSettings = ThrottlingSettings()
    )

    data class NativeSettings(
        val torConnected: Boolean = true,
        val torDisconnected: Boolean = true,
        val proxyError: Boolean = true,
        val protectionStatusChange: Boolean = true,
        val suspiciousActivity: Boolean = true,
        val extensionWarning: Boolean = true
    )

    data class OverlaySettings(
        val securityWarnings: Boolean = true,
        val rpcBlocked: Boolean = true,
        val suspiciousExtension: Boolean = true,
        val unprotectedWarning: Boolean = true
    )

    data class ThrottlingSettings(
        val rpcActivityCooldownMs: Long = 30_000,
        val proxyErrorCooldownMs: Long = 60_000,
        val maxPerMinute: Int = 5
    )

    // ── Notification Data ────────────────────────────────────

    data class Notification(
        val type: NotificationType,
        val title: String,
        val message: String,
        val priority: Int = type.priority,
        val actions: List<NotificationAction> = emptyList(),
        val timestamp: Long = System.currentTimeMillis()
    )

    data class NotificationAction(
        val label: String,
        val action: String
    )

    data class NotifyResult(
        val throttled: Boolean = false,
        val nativeSent: Boolean = false,
        val overlaySent: Boolean = false
    )

    // ── Callback Interfaces ──────────────────────────────────

    fun interface NativeNotificationSender {
        fun send(notification: Notification): Boolean
    }

    fun interface OverlayNotificationSender {
        fun send(notification: Notification): Boolean
    }

    fun interface ActionHandler {
        fun handle(notificationId: String, action: String)
    }

    // ── State ────────────────────────────────────────────────

    private val lastNotifications = mutableMapOf<NotificationType, Long>()
    private val recentTimestamps = mutableListOf<Long>()
    private var nativeSender: NativeNotificationSender? = null
    private var overlaySender: OverlayNotificationSender? = null
    private var actionHandler: ActionHandler? = null

    // ── Configuration ────────────────────────────────────────

    fun setNativeSender(sender: NativeNotificationSender) {
        nativeSender = sender
    }

    fun setOverlaySender(sender: OverlayNotificationSender) {
        overlaySender = sender
    }

    fun setActionHandler(handler: ActionHandler) {
        actionHandler = handler
    }

    fun updateSettings(newSettings: NotificationSettings) {
        settings = newSettings
    }

    fun getSettings(): NotificationSettings = settings

    // ── Core ─────────────────────────────────────────────────

    /**
     * Send a notification through configured channels.
     * Respects throttling and per-type/per-channel settings.
     */
    fun notify(notification: Notification): NotifyResult {
        // Check throttling
        if (shouldThrottle(notification.type)) {
            return NotifyResult(throttled = true)
        }

        var nativeSent = false
        var overlaySent = false

        // Native channel
        if (isEnabled(notification.type, Channel.NATIVE)) {
            nativeSent = nativeSender?.send(notification) ?: false
        }

        // Overlay channel
        if (isEnabled(notification.type, Channel.OVERLAY)) {
            overlaySent = overlaySender?.send(notification) ?: false
        }

        // Record
        if (nativeSent || overlaySent) {
            recordNotification(notification.type)
        }

        return NotifyResult(
            throttled = false,
            nativeSent = nativeSent,
            overlaySent = overlaySent
        )
    }

    /** Handle a notification action callback */
    fun handleAction(notificationId: String, action: String) {
        actionHandler?.handle(notificationId, action)
    }

    // ── Helpers ──────────────────────────────────────────────

    private enum class Channel { NATIVE, OVERLAY }

    private fun shouldThrottle(type: NotificationType): Boolean {
        val now = System.currentTimeMillis()

        // Clean old entries
        recentTimestamps.removeAll { now - it >= 60_000 }

        // Global rate limit
        if (recentTimestamps.size >= settings.throttling.maxPerMinute) {
            return true
        }

        // Per-type cooldown
        val lastTime = lastNotifications[type] ?: 0
        val cooldown = when (type) {
            NotificationType.SUSPICIOUS_RPC,
            NotificationType.RPC_BLOCKED -> settings.throttling.rpcActivityCooldownMs
            NotificationType.PROXY_ERROR -> settings.throttling.proxyErrorCooldownMs
            else -> 0L
        }

        return cooldown > 0 && now - lastTime < cooldown
    }

    private fun recordNotification(type: NotificationType) {
        val now = System.currentTimeMillis()
        lastNotifications[type] = now
        recentTimestamps.add(now)
    }

    private fun isEnabled(type: NotificationType, channel: Channel): Boolean {
        return when (channel) {
            Channel.NATIVE -> {
                if (!settings.nativeEnabled || !type.supportsNative) return false
                when (type) {
                    NotificationType.TOR_CONNECTED -> settings.native.torConnected
                    NotificationType.TOR_DISCONNECTED -> settings.native.torDisconnected
                    NotificationType.PROXY_ERROR -> settings.native.proxyError
                    NotificationType.PROTECTION_ON,
                    NotificationType.PROTECTION_OFF -> settings.native.protectionStatusChange
                    NotificationType.SUSPICIOUS_RPC -> settings.native.suspiciousActivity
                    NotificationType.EXT_WARNING -> settings.native.extensionWarning
                    else -> true
                }
            }
            Channel.OVERLAY -> {
                if (!settings.overlayEnabled || !type.supportsOverlay) return false
                when (type) {
                    NotificationType.PROXY_ERROR,
                    NotificationType.PROTECTION_OFF,
                    NotificationType.SUSPICIOUS_RPC -> settings.overlay.securityWarnings
                    NotificationType.RPC_BLOCKED -> settings.overlay.rpcBlocked
                    NotificationType.EXT_WARNING -> settings.overlay.suspiciousExtension
                    NotificationType.UNPROTECTED_DAPP -> settings.overlay.unprotectedWarning
                    else -> true
                }
            }
        }
    }
}

File: sdk/swift/Sources/PrivacyRPCSDK/NotificationHub.swift
import Foundation

/// Notification Hub for PrivacyRPC
///
/// Central notification management matching the Chrome extension's
/// notification system with throttling, type-based filtering,
/// and multi-channel support.
public class NotificationHub {

    // MARK: - Notification Types

    public enum NotificationType: String, CaseIterable {
        case torConnected = "TOR_CONNECTED"
        case torDisconnected = "TOR_DISCONNECTED"
        case proxyError = "PROXY_ERROR"
        case protectionOn = "PROTECTION_ON"
        case protectionOff = "PROTECTION_OFF"
        case suspiciousRpc = "SUSPICIOUS_RPC"
        case extWarning = "EXT_WARNING"
        case unprotectedDApp = "UNPROTECTED_DAPP"
        case rpcBlocked = "RPC_BLOCKED"

        public var priority: Int {
            switch self {
            case .proxyError, .suspiciousRpc: return 100
            case .torConnected, .torDisconnected, .protectionOff, .extWarning, .rpcBlocked: return 80
            case .protectionOn, .unprotectedDApp: return 50
            }
        }

        public var supportsNative: Bool {
            switch self {
            case .unprotectedDApp, .rpcBlocked: return false
            default: return true
            }
        }

        public var supportsOverlay: Bool {
            switch self {
            case .torConnected, .torDisconnected, .protectionOn: return false
            default: return true
            }
        }
    }

    // MARK: - Settings

    public struct Settings {
        public var nativeEnabled: Bool
        public var overlayEnabled: Bool
        public var native: NativeSettings
        public var overlay: OverlaySettings
        public var throttling: ThrottlingSettings

        public init(
            nativeEnabled: Bool = true,
            overlayEnabled: Bool = true,
            native: NativeSettings = NativeSettings(),
            overlay: OverlaySettings = OverlaySettings(),
            throttling: ThrottlingSettings = ThrottlingSettings()
        ) {
            self.nativeEnabled = nativeEnabled
            self.overlayEnabled = overlayEnabled
            self.native = native
            self.overlay = overlay
            self.throttling = throttling
        }
    }

    public struct NativeSettings {
        public var torConnected = true
        public var torDisconnected = true
        public var proxyError = true
        public var protectionStatusChange = true
        public var suspiciousActivity = true
        public var extensionWarning = true

        public init() {}
    }

    public struct OverlaySettings {
        public var securityWarnings = true
        public var rpcBlocked = true
        public var suspiciousExtension = true
        public var unprotectedWarning = true

        public init() {}
    }

    public struct ThrottlingSettings {
        public var rpcActivityCooldown: TimeInterval = 30
        public var proxyErrorCooldown: TimeInterval = 60
        public var maxPerMinute: Int = 5

        public init() {}
    }

    // MARK: - Notification

    public struct Notification {
        public let type: NotificationType
        public let title: String
        public let message: String
        public let priority: Int
        public let actions: [NotificationAction]
        public let timestamp: Date

        public init(
            type: NotificationType,
            title: String,
            message: String,
            priority: Int? = nil,
            actions: [NotificationAction] = [],
            timestamp: Date = Date()
        ) {
            self.type = type
            self.title = title
            self.message = message
            self.priority = priority ?? type.priority
            self.actions = actions
            self.timestamp = timestamp
        }
    }

    public struct NotificationAction {
        public let label: String
        public let action: String

        public init(label: String, action: String) {
            self.label = label
            self.action = action
        }
    }

    public struct NotifyResult {
        public let throttled: Bool
        public let nativeSent: Bool
        public let overlaySent: Bool
    }

    // MARK: - Callbacks

    public typealias NativeSender = (Notification) -> Bool
    public typealias OverlaySender = (Notification) -> Bool
    public typealias ActionHandler = (String, String) -> Void

    // MARK: - State

    public var settings: Settings

    private var lastNotifications: [NotificationType: Date] = [:]
    private var recentTimestamps: [Date] = []
    private var nativeSender: NativeSender?
    private var overlaySender: OverlaySender?
    private var actionHandler: ActionHandler?

    public init(settings: Settings = Settings()) {
        self.settings = settings
    }

    // MARK: - Configuration

    public func setNativeSender(_ sender: @escaping NativeSender) {
        nativeSender = sender
    }

    public func setOverlaySender(_ sender: @escaping OverlaySender) {
        overlaySender = sender
    }

    public func setActionHandler(_ handler: @escaping ActionHandler) {
        actionHandler = handler
    }

    // MARK: - Core

    /// Send a notification through configured channels
    @discardableResult
    public func notify(_ notification: Notification) -> NotifyResult {
        if shouldThrottle(notification.type) {
            return NotifyResult(throttled: true, nativeSent: false, overlaySent: false)
        }

        var nativeSent = false
        var overlaySent = false

        if isEnabled(notification.type, channel: .native) {
            nativeSent = nativeSender?(notification) ?? false
        }

        if isEnabled(notification.type, channel: .overlay) {
            overlaySent = overlaySender?(notification) ?? false
        }

        if nativeSent || overlaySent {
            recordNotification(notification.type)
        }

        return NotifyResult(throttled: false, nativeSent: nativeSent, overlaySent: overlaySent)
    }

    /// Handle a notification action
    public func handleAction(notificationId: String, action: String) {
        actionHandler?(notificationId, action)
    }

    // MARK: - Private

    private enum Channel { case native, overlay }

    private func shouldThrottle(_ type: NotificationType) -> Bool {
        let now = Date()

        // Clean old entries
        recentTimestamps.removeAll { now.timeIntervalSince($0) >= 60 }

        // Global rate limit
        if recentTimestamps.count >= settings.throttling.maxPerMinute {
            return true
        }

        // Per-type cooldown
        if let lastTime = lastNotifications[type] {
            let cooldown: TimeInterval
            switch type {
            case .suspiciousRpc, .rpcBlocked:
                cooldown = settings.throttling.rpcActivityCooldown
            case .proxyError:
                cooldown = settings.throttling.proxyErrorCooldown
            default:
                cooldown = 0
            }

            if cooldown > 0 && now.timeIntervalSince(lastTime) < cooldown {
                return true
            }
        }

        return false
    }

    private func recordNotification(_ type: NotificationType) {
        let now = Date()
        lastNotifications[type] = now
        recentTimestamps.append(now)
    }

    private func isEnabled(_ type: NotificationType, channel: Channel) -> Bool {
        switch channel {
        case .native:
            guard settings.nativeEnabled, type.supportsNative else { return false }
            switch type {
            case .torConnected: return settings.native.torConnected
            case .torDisconnected: return settings.native.torDisconnected
            case .proxyError: return settings.native.proxyError
            case .protectionOn, .protectionOff: return settings.native.protectionStatusChange
            case .suspiciousRpc: return settings.native.suspiciousActivity
            case .extWarning: return settings.native.extensionWarning
            default: return true
            }

        case .overlay:
            guard settings.overlayEnabled, type.supportsOverlay else { return false }
            switch type {
            case .proxyError, .protectionOff, .suspiciousRpc: return settings.overlay.securityWarnings
            case .rpcBlocked: return settings.overlay.rpcBlocked
            case .extWarning: return settings.overlay.suspiciousExtension
            case .unprotectedDApp: return settings.overlay.unprotectedWarning
            default: return true
            }
        }
    }
}
